File: contrib/DATA_Analysis/Vespucci/vespucci/pre_processing_core.py
from distutils.log import error
import numpy as np
import scipy.signal as sig
import matplotlib.pyplot as plt
import cmsisdsp as dsp
import cmsisdsp.fixedpoint as f
import scipy.signal as sig
from cmsisdsp.datatype import F32


TRIANGULAR_FILTERS_BANK_SIZE = 128

'''
        BASIC MATH FUNCTIONS
'''
def complex_abs( real,  compl):
    return np.sqrt(real*real + compl*compl)

def dot_product(in_1, in_2):
    mult = dsp.arm_mult_f32(in_1,in_2)
    sum = 0
    for i in range(len(mult)):
        sum = sum + mult[i]
    return sum

def hz_to_mel(f_hz):
    return(2595*np.log10(1+f_hz/700))

def mel_to_hz(f_mel):
    return(700*(10**(f_mel/2505)-1))


'''
        SIGNAL PROCESSING FUNCTIONS
'''
def axis_selection(data_in_df , accelerometer_axis):
    signal = np.array( data_in_df['A_'+accelerometer_axis+' [g]'] )
    return signal

def mean_removal(data_in):
    mean = np.mean(data_in)
    data_out = data_in-mean
    return data_out

def signal_normalization(data_in, peak_to_peak, offset):
    max = max(data_in)
    min = min(data_in)
    data_out = ((data_in-min)/(max-min))*peak_to_peak - ((peak_to_peak/2)+offset)
    return data_out

def hanning(data_in):
    multiplier = sig.hanning(len(data_in))
    data_out = data_in * multiplier
    return data_out

def fft_(data_in, signal_windowing):
    number_of_samples = len(data_in)
    rfftf32=dsp.arm_rfft_fast_instance_f32()
    fft_status=dsp.arm_rfft_fast_init_f32(rfftf32,number_of_samples)
    if (signal_windowing == 'HANNING'):
        data_in = hanning(data_in)

    if (fft_status == 0):
        fft_out_buffer = dsp.arm_rfft_fast_f32(rfftf32,data_in,0)
        fft = np.zeros(int(number_of_samples/2))
        freq_point = 0
        for ii in range(0, number_of_samples ,2):
            fft[freq_point] = complex_abs( fft_out_buffer[ii],  fft_out_buffer[ii+1])
            if fft[freq_point] < 1e-3:
                fft[freq_point] = 1e-3
            freq_point = freq_point+1
        
        return fft
    else:
        print('ERROR: INPUT DATA DIMENSIONS NOT SUPPORTED!!!!')
        return error

def triangular_filters_init(number_of_samples, sampling_frequency, triangular_filter_scale):
    TRIANGULAR_FILTERS_BANK_FRACTION = 0.45
    f_min = 0
    f_max = sampling_frequency * TRIANGULAR_FILTERS_BANK_FRACTION

    if (triangular_filter_scale == 'TRIANGULAR_FILTERS_SCALE_MEL'):
        low_freq = hz_to_mel(f_min)
        high_freq = hz_to_mel(f_max)
    else:
        low_freq = f_min
        high_freq = f_max

    d_hz_points = (high_freq-low_freq)/(TRIANGULAR_FILTERS_BANK_SIZE+2)
    hz_points = np.zeros(TRIANGULAR_FILTERS_BANK_SIZE+2)
    bin_sep=sampling_frequency/number_of_samples
    bin = np.zeros(TRIANGULAR_FILTERS_BANK_SIZE+2)

    for ii in range(TRIANGULAR_FILTERS_BANK_SIZE+2):
        if (triangular_filter_scale == 'TRIANGULAR_FILTERS_SCALE_MEL'):
            hz_points[ii] = mel_to_hz((low_freq + ii * d_hz_points))
        else:
            hz_points[ii] = low_freq + ii * d_hz_points
        
        bin[ii] =  np.round(hz_points[ii]/bin_sep)
    
    return bin
        
def triangular_filters_convolution(data_in, bin ):
    # fft or spectrum as input
    number_of_samples = len(data_in)
    fbank=np.zeros((TRIANGULAR_FILTERS_BANK_SIZE,int(np.floor(number_of_samples))))
    for m in range(1,TRIANGULAR_FILTERS_BANK_SIZE+1):
        f_m_minus=int(bin[m-1])
        f_m=int(bin[m])
        f_m_plus=int(bin[m+1])
        for k in range(f_m_minus, f_m):
            fbank[m-1,k]=(k-f_m_minus)/(f_m-f_m_minus)
        for k in range(f_m,f_m_plus):
            fbank[m-1,k]=(f_m_plus-k)/(f_m_plus-f_m)
    
    mel_spectrum = np.zeros(128)
    for ii in range(fbank.shape[0]):
        mel_spectrum[ii] = dot_product(fbank[ii,:], data_in)  
    
    return mel_spectrum
  
def mfcc_(data_in, bin, signal_windowing):
    dct_support_vector= np.zeros(int(TRIANGULAR_FILTERS_BANK_SIZE * 2))
    spectrum  = fft_(data_in, signal_windowing)
    dct4f32=dsp.arm_dct4_instance_f32()
    cfftradix4f32=dsp.arm_cfft_radix4_instance_f32()
    rfftf32=dsp.arm_rfft_fast_instance_f32()
    dct_status=dsp.arm_dct4_init_f32(dct4f32,rfftf32,cfftradix4f32,TRIANGULAR_FILTERS_BANK_SIZE,int(TRIANGULAR_FILTERS_BANK_SIZE/2),0.125)
   
    if dct_status == 0:
        mel_spectrum = triangular_filters_convolution(spectrum , bin )
        mel_spectrum = 20* np.log10(mel_spectrum)
        mel_frequency_cepstral_coef = dsp.arm_dct4_f32(dct4f32,dct_support_vector,mel_spectrum)
        return mel_frequency_cepstral_coef
    else:
        return error

File: firmware/ai_vibration_analysis_stwin_box_bm/Core/pre_processing_library/Src/pre_processing_core.c
/*
 * pre_processing_core.c
 *
 *  Created on: Feb 16, 2022
 *      Author: leonardoiacussi
 */

#include "pre_processing_core.h"
#include "arm_math.h"

/*----------------------------------------------------------------------------*/
/* Support functions          			      						          */
/*----------------------------------------------------------------------------*/

// calculation of the complex absolute value
float complex_abs(float real, float compl) {
	return sqrtf(real*real+compl*compl);
}

// Dot product calculation using CMSIS DSP library
float32_t dot_product(float32_t* in_1, uint32_t data_in1_size, float32_t* in_2, uint32_t data_in2_size) {

	float32_t multOutput[data_in1_size];
	float32_t dot_out=0;
	arm_mult_f32(in_1, in_2, multOutput, data_in1_size);
	for (int i=0;  i<data_in1_size;   i++){
		arm_add_f32(&dot_out, &multOutput[i], &dot_out, 1);
	}
	return dot_out;
}

// conversion between Hz and Mel
float hz_to_mel(float f_Hz_in) {
	return 2595*log10(1+f_Hz_in/700);
}

float mel_to_hz(float f_Mel_in) {
	return (700*(pow(10,f_Mel_in/2505)-1));
}

/*----------------------------------------------------------------------------*/
/* Signal pre-processing :  Axis selection        						      */
/*----------------------------------------------------------------------------*/

void axis_selection( tridimensional_data_t * data_in, uint32_t data_in_size, float32_t * data_out, uint32_t data_out_size, axis_t axis) {
	if (axis == X)
	{
		for (int i=0; i<data_out_size ; i++){
	    	data_out[i]=data_in[i].x;
	    }
	}
	else if (axis == Y)
	{
		for (int i=0; i<data_out_size ; i++){
	    	data_out[i]=data_in[i].y;
	    }
	}
	else if (axis == Z)
	{
		for (int i=0; i<data_out_size ; i++){
	    	data_out[i]=data_in[i].z;
	    }
	}
}

/*----------------------------------------------------------------------------*/
/* Signal pre-processing :  Remove mean from signal     			          */
/*----------------------------------------------------------------------------*/

void mean_removal (float32_t * data_in, uint32_t data_in_size, float32_t * data_out, uint32_t data_out_size) {

	static float32_t mean;

	// calculation of the signal mean value
	arm_mean_f32(data_in, data_in_size, &mean);

	// subtract the mean value from the signal in order to remove gravity effect on the accelerometric data
	for (int i=0 ; i < data_out_size ; i++){
		data_out[i] = data_in[i] - mean;
	}
}

/*----------------------------------------------------------------------------*/
/* Signal pre-processing :  Signal normalization between -1 and 1             */
/*----------------------------------------------------------------------------*/

void signal_normalization(float32_t *data_in, uint32_t data_in_size, float32_t * data_out, uint32_t data_out_size) {

	float32_t max, min;
	uint32_t maxpos, minpos;

	arm_max_f32(data_in,data_in_size,&max,&maxpos);
	arm_min_f32(data_in,data_in_size,&min,&minpos);

	for(int i=0; i<data_out_size ; i++){
		data_out[i] = ((data_in[i] - min)/(max - min))*2 - 1;
	}
}

/*----------------------------------------------------------------------------*/
/* Signal pre-processing :  Multiply the Hanning window to the input signal   */
/*----------------------------------------------------------------------------*/

void hanning(float32_t *data_in, uint32_t data_in_size, float32_t * data_out, uint32_t data_out_size) {
	for (int i = 0; i < data_in_size; i++) {
		float32_t multiplier = 0.5 * (1 - arm_cos_f32((float32_t) 2*PI*i/(data_out_size-1)));
		data_out[i] = multiplier * data_in[i];
	}
}


/*----------------------------------------------------------------------------*/
/* Signal pre-processing :  FFT Calculation using CMSIS DSP Library           */
/*----------------------------------------------------------------------------*/

void fft(float32_t *data_in, uint32_t data_in_size, float32_t * data_out, uint32_t data_out_size, arm_rfft_fast_instance_f32 * fft_handler, signal_windowing_t signal_windowing) {

	float32_t fft_out_buf[data_in_size];

	switch (signal_windowing){
		case HANNING:
		{
			hanning(data_in, data_in_size, data_in, data_in_size);
			break;
		}
		default: {
			// no window applied (rectangular case)
			break;
		}
	}

	arm_rfft_fast_f32(fft_handler, (float32_t*)  data_in, (float32_t*) fft_out_buf, 0);

	int freqpoint = 0;

	for (int i=0;   i<data_in_size   ; i=i+2) {
	  data_out[freqpoint] =(complex_abs(fft_out_buf[i], fft_out_buf[i+1])); //(sqrt(data_in_size));
	  //data_out[freqpoint] = 2 * data_out[freqpoint]*data_out[freqpoint] /(float32_t)ISM330DHCX_ODR;

	  if(data_out[freqpoint] < 1e-3){
		  data_out[freqpoint] = 1e-3;
	  }
	  freqpoint++;
	}
}

/*----------------------------------------------------------------------------*/
/* Signal pre-processing :  Calculation of the triangular filters bank        */
/*----------------------------------------------------------------------------*/

void triangular_filters_bank(uint32_t number_of_samples, float32_t odr, triangular_filters_scale_t triangular_filters_scale, uint32_t* bin) {

	float32_t low_freq;
	float32_t high_freq;
	float32_t Hz_points[TRIANGULAR_FILTERS_BANK_SIZE+2];
	float32_t d_hz_points;
	float32_t bin_sep;
	float32_t f_max;
	float32_t f_min = 0.0;

	f_max = odr * TRIANGULAR_FILTERS_BANK_FRACTION;

	if (triangular_filters_scale == TRIANGULAR_FILTERS_SCALE_MEL) {
	    low_freq = hz_to_mel(f_min);
	    high_freq = hz_to_mel(f_max);
	} else {
        low_freq = f_min;
	    high_freq = f_max;
	}

	d_hz_points = (high_freq - low_freq) / (float32_t) (TRIANGULAR_FILTERS_BANK_SIZE+2);

	bin_sep = odr / (float32_t) number_of_samples;

	for (int i=0; i < TRIANGULAR_FILTERS_BANK_SIZE+2; i++){
		Hz_points[i] = (float32_t) (low_freq + i * d_hz_points);
		bin[i] = round((Hz_points[i] / bin_sep));
	}
}

/*----------------------------------------------------------------------------*/
/* Signal pre-processing :  Convolution of the triangular filters bank        */
/*----------------------------------------------------------------------------*/

void triangular_filters_convolution(float32_t * data_in, uint32_t data_in_size, float32_t * data_out,  uint32_t data_out_size, uint32_t * bin) {

	int f_m_minus=0, f_m=0, f_m_plus=0;
	float32_t in_vector[data_in_size];

	for (int m=1; m < data_out_size+1; m++){
		f_m_minus = bin[m-1];
		f_m = bin[m];
		f_m_plus = bin[m+1];

		for (int i=0; i<data_in_size   ;i++){
			in_vector[i] = (float32_t) 0.0;
		}
		for (int k=f_m_minus;k < f_m;k++){
			in_vector[k] = (((float32_t)k-f_m_minus)/(f_m-f_m_minus));
		}
		for (int j=f_m; j<f_m_plus;j++){
			in_vector[j] = (float32_t)((f_m_plus-(float32_t)j)/(f_m_plus-f_m));
		}
		data_out[m-1] = dot_product(in_vector, data_in_size,data_in, data_in_size);
	}
}

/*----------------------------------------------------------------------------*/
/* Signal pre-processing :  MFCC		        					          */
/*----------------------------------------------------------------------------*/

void mfcc(float32_t * data_in, uint32_t data_in_size, float32_t * data_out, uint32_t data_out_size, uint32_t * bin, arm_dct4_instance_f32 * dct4f32, arm_rfft_fast_instance_f32 * fft_handler,  signal_windowing_t signal_windowing) {

	float32_t dct_status[data_out_size*2];
	float32_t fft_out[data_in_size/2];

	fft(data_in, data_in_size, fft_out, data_in_size/2, fft_handler, signal_windowing);

	triangular_filters_convolution(fft_out, data_in_size/2, data_out, data_out_size,bin);
	for (int i = 0; i<data_out_size; i++) {
		data_out[i] = 20*log10(data_out[i]);
	}

	// apply the direct cosine transform in order to have the MFCC
	arm_dct4_f32(dct4f32, dct_status, data_out);
}
